// @copilotkit/react-core/dist/chunk-H3FQWB4B.mjs
import {
  StandardCopilotApiConfig
} from "./chunk-F2JIAPZQ.mjs";
import {
  use_flat_category_store_default
} from "./chunk-YULKJPY3.mjs";
import {
  use_tree_default
} from "./chunk-VUY2K2DI.mjs";
import {
  CopilotContext
} from "./chunk-KIKHLN4X.mjs";
import {
  __async,
  __objRest,
  __spreadProps,
  __spreadValues
} from "./chunk-MRXNTQOX.mjs";

// src/components/copilot-provider/copilotkit.tsx
import { useCallback, useRef, useState } from "react";
import { actionToChatCompletionFunction } from "@copilotkit/shared";
import { jsx } from "react/jsx-runtime";
function CopilotKit(_a) {
  var _b = _a, { children } = _b, props = __objRest(_b, ["children"]);
  const [entryPoints, setEntryPoints] = useState({});
  const chatComponentsCache = useRef({});
  const { addElement, removeElement, printTree } = use_tree_default();
  const {
    addElement: addDocument,
    removeElement: removeDocument,
    allElements: allDocuments
  } = use_flat_category_store_default();
  const setEntryPoint = useCallback((id, entryPoint) => {
    setEntryPoints((prevPoints) => {
      return __spreadProps(__spreadValues({}, prevPoints), {
        [id]: entryPoint
      });
    });
  }, []);
  const removeEntryPoint = useCallback((id) => {
    setEntryPoints((prevPoints) => {
      const newPoints = __spreadValues({}, prevPoints);
      delete newPoints[id];
      return newPoints;
    });
  }, []);
  const getContextString = useCallback(
    (documents, categories) => {
      const documentsString = documents.map((document) => {
        return `${document.name} (${document.sourceApplication}):
${document.getContents()}`;
      }).join("\n\n");
      const nonDocumentStrings = printTree(categories);
      return `${documentsString}

${nonDocumentStrings}`;
    },
    [printTree]
  );
  const addContext = useCallback(
    (context, parentId, categories = defaultCopilotContextCategories) => {
      return addElement(context, categories, parentId);
    },
    [addElement]
  );
  const removeContext = useCallback(
    (id) => {
      removeElement(id);
    },
    [removeElement]
  );
  const getChatCompletionFunctionDescriptions = useCallback(
    (customEntryPoints) => {
      return entryPointsToChatCompletionFunctions(Object.values(customEntryPoints || entryPoints));
    },
    [entryPoints]
  );
  const getFunctionCallHandler = useCallback(
    (customEntryPoints) => {
      return entryPointsToFunctionCallHandler(Object.values(customEntryPoints || entryPoints));
    },
    [entryPoints]
  );
  const getDocumentsContext = useCallback(
    (categories) => {
      return allDocuments(categories);
    },
    [allDocuments]
  );
  const addDocumentContext = useCallback(
    (documentPointer, categories = defaultCopilotContextCategories) => {
      return addDocument(documentPointer, categories);
    },
    [addDocument]
  );
  const removeDocumentContext = useCallback(
    (documentId) => {
      removeDocument(documentId);
    },
    [removeDocument]
  );
  const copilotApiConfig = new StandardCopilotApiConfig(
    props.url,
    `${props.url}/v2`,
    props.headers || {},
    __spreadValues(__spreadValues({}, props.body), props.backendOnlyProps)
  );
  return /* @__PURE__ */ jsx(
    CopilotContext.Provider,
    {
      value: {
        entryPoints,
        chatComponentsCache,
        getChatCompletionFunctionDescriptions,
        getFunctionCallHandler,
        setEntryPoint,
        removeEntryPoint,
        getContextString,
        addContext,
        removeContext,
        getDocumentsContext,
        addDocumentContext,
        removeDocumentContext,
        copilotApiConfig
      },
      children
    }
  );
}
var defaultCopilotContextCategories = ["global"];
function entryPointsToChatCompletionFunctions(actions) {
  return actions.map(actionToChatCompletionFunction);
}
function entryPointsToFunctionCallHandler(actions) {
  return (chatMessages, functionCall) => __async(this, null, function* () {
    let actionsByFunctionName = {};
    for (let action2 of actions) {
      actionsByFunctionName[action2.name] = action2;
    }
    const action = actionsByFunctionName[functionCall.name || ""];
    if (action) {
      let functionCallArguments = [];
      if (functionCall.arguments) {
        functionCallArguments = JSON.parse(functionCall.arguments);
      }
      return yield action.handler(functionCallArguments);
    }
  });
}

export {
  CopilotKit,
  defaultCopilotContextCategories
};
