// @copilotkit/backend/dist/chunk-LQ4ABAUZ.mjs
import {
  remoteChainToAction
} from "./chunk-7ZFVFHJF.mjs";
import {
  copilotkitStreamInterceptor
} from "./chunk-JIL2G6XG.mjs";

// src/lib/copilotkit-backend-implementation.ts
import {
  EXCLUDE_FROM_FORWARD_PROPS_KEYS,
  actionToChatCompletionFunction,
  annotatedFunctionToAction
} from "@copilotkit/shared";
var CopilotBackendImplementation = class {
  constructor(params) {
    this.actions = [];
    this.langserve = [];
    this.debug = false;
    for (const action of (params == null ? void 0 : params.actions) || []) {
      if ("argumentAnnotations" in action) {
        this.actions.push(annotatedFunctionToAction(action));
      } else {
        this.actions.push(action);
      }
    }
    for (const chain of (params == null ? void 0 : params.langserve) || []) {
      this.langserve.push(remoteChainToAction(chain));
    }
    this.debug = (params == null ? void 0 : params.debug) || false;
  }
  addAction(action) {
    this.removeAction(action.name);
    if ("argumentAnnotations" in action) {
      this.actions.push(annotatedFunctionToAction(action));
    } else {
      this.actions.push(action);
    }
  }
  removeAction(actionName) {
    this.actions = this.actions.filter((f) => f.name !== actionName);
  }
  removeBackendOnlyProps(forwardedProps) {
    const backendOnlyPropsKeys = forwardedProps[EXCLUDE_FROM_FORWARD_PROPS_KEYS];
    if (Array.isArray(backendOnlyPropsKeys)) {
      backendOnlyPropsKeys.forEach((key) => {
        const success2 = Reflect.deleteProperty(forwardedProps, key);
        if (!success2) {
          console.error(`Failed to delete property ${key}`);
        }
      });
      const success = Reflect.deleteProperty(forwardedProps, EXCLUDE_FROM_FORWARD_PROPS_KEYS);
      if (!success) {
        console.error(`Failed to delete EXCLUDE_FROM_FORWARD_PROPS_KEYS`);
      }
    } else if (backendOnlyPropsKeys) {
      console.error("backendOnlyPropsKeys is not an array");
    }
  }
  async getResponse(forwardedProps, serviceAdapter) {
    this.removeBackendOnlyProps(forwardedProps);
    const langserveFunctions = [];
    for (const chainPromise of this.langserve) {
      try {
        const chain = await chainPromise;
        langserveFunctions.push(chain);
      } catch (error) {
        console.error("Error loading langserve chain:", error);
      }
    }
    let mergedTools = mergeServerSideTools(
      this.actions.map(actionToChatCompletionFunction),
      langserveFunctions.map(actionToChatCompletionFunction)
    );
    mergedTools = mergeServerSideTools(mergedTools, forwardedProps.tools);
    try {
      const result = await serviceAdapter.getResponse({
        ...forwardedProps,
        tools: mergedTools
      });
      const stream = copilotkitStreamInterceptor(
        result.stream,
        [...this.actions, ...langserveFunctions],
        this.debug
      );
      return { stream, headers: result.headers };
    } catch (error) {
      console.error("Error getting response:", error);
      throw error;
    }
  }
  async response(req, serviceAdapter) {
    try {
      const response = await this.getResponse(await req.json(), serviceAdapter);
      return new Response(response.stream, { headers: response.headers });
    } catch (error) {
      return new Response("", { status: 500, statusText: error.message });
    }
  }
  async streamHttpServerResponse(req, res, serviceAdapter, headers) {
    const bodyParser = new Promise((resolve, reject) => {
      if ("body" in req) {
        resolve(req.body);
        return;
      }
      let body = "";
      req.on("data", (chunk) => body += chunk.toString());
      req.on("end", () => {
        try {
          resolve(JSON.parse(body));
        } catch (error) {
          reject(error);
        }
      });
    });
    const forwardedProps = await bodyParser;
    const response = await this.getResponse(forwardedProps, serviceAdapter);
    const mergedHeaders = { ...headers, ...response.headers };
    res.writeHead(200, mergedHeaders);
    const stream = response.stream;
    const reader = stream.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        res.end();
        break;
      } else {
        res.write(new TextDecoder().decode(value));
      }
    }
  }
};
function mergeServerSideTools(serverTools, clientTools) {
  let allTools = serverTools.slice();
  const serverToolsNames = serverTools.map((tool) => tool.function.name);
  if (clientTools) {
    allTools = allTools.concat(
      // filter out any client functions that are already defined on the server
      clientTools.filter((tool) => !serverToolsNames.includes(tool.function.name))
    );
  }
  return allTools;
}

export {
  CopilotBackendImplementation,
  mergeServerSideTools
};
